File: php/tests/HqcTest.php
<?php

declare(strict_types=1);

namespace PQC\Tests;

use PHPUnit\Framework\TestCase;
use PQC\Hqc\GF256;
use PQC\Hqc\GF2;
use PQC\Hqc\ReedMuller;
use PQC\Hqc\ReedSolomon;
use PQC\Hqc\TensorCode;
use PQC\Hqc\HqcParams;
use PQC\Hqc\Hqc;

class HqcTest extends TestCase
{
    public function testGF256Tables(): void
    {
        GF256::init();
        for ($i = 1; $i < 256; $i++) {
            $this->assertSame($i, GF256::mul(1, $i), "1 * $i should be $i");
        }
        for ($i = 1; $i < 256; $i++) {
            $this->assertSame(1, GF256::mul($i, GF256::inv($i)), "$i * inv($i) should be 1");
        }
    }

    public function testGF2VectOps(): void
    {
        $a = gmp_init('AAAA', 16);
        $b = gmp_init('5555', 16);
        $c = GF2::add($a, $b);
        $this->assertSame('ffff', gmp_strval($c, 16));
    }

    public function testRMEncodeDecodeRoundtrip(): void
    {
        for ($msg = 0; $msg < 256; $msg++) {
            foreach ([3, 5] as $mult) {
                $n2 = $mult * 128;
                $cw = GF2::zero();
                ReedMuller::encodeInto($cw, $msg, 0, $mult);
                $decoded = ReedMuller::decode($cw, $n2, $mult);
                $this->assertSame($msg, $decoded, "RM mult=$mult msg=$msg: got $decoded");
            }
        }
    }

    /**
     * @dataProvider paramProvider
     */
    public function testRSEncodeDecodeRoundtrip(HqcParams $p): void
    {
        $msg = [];
        for ($i = 0; $i < $p->k; $i++) {
            $msg[] = ($i + 1) & 0xFF;
        }
        $cw = ReedSolomon::encode($msg, $p);
        [$decoded, $ok] = ReedSolomon::decode($cw, $p);
        $this->assertTrue($ok, 'decode failed on clean codeword');
        $this->assertSame($msg, $decoded, 'roundtrip mismatch');
    }

    /**
     * @dataProvider paramProvider
     */
    public function testRSDecodeWithErrors(HqcParams $p): void
    {
        $msg = [];
        for ($i = 0; $i < $p->k; $i++) {
            $msg[] = ($i * 3 + 7) & 0xFF;
        }
        $cw = ReedSolomon::encode($msg, $p);

        // Inject delta errors
        for ($i = 0; $i < $p->delta; $i++) {
            $cw[$i] ^= ($i + 1) & 0xFF;
        }

        [$decoded, $ok] = ReedSolomon::decode($cw, $p);
        $this->assertTrue($ok, 'decode failed with correctable errors');
        $this->assertSame($msg, $decoded, 'decode mismatch after error correction');
    }

    /**
     * @dataProvider paramProvider
     */
    public function testTensorEncodeDecodeRoundtrip(HqcParams $p): void
    {
        $msg = [];
        for ($i = 0; $i < $p->k; $i++) {
            $msg[] = ($i + 42) & 0xFF;
        }
        $encoded = TensorCode::encode($msg, $p);
        [$decoded, $ok] = TensorCode::decode($encoded, $p);
        $this->assertTrue($ok, 'tensor decode failed');
        $this->assertSame($msg, $decoded, 'tensor roundtrip mismatch');
    }

    /**
     * @dataProvider paramProvider
     */
    public function testKEMRoundtrip(HqcParams $p): void
    {
        [$pk, $sk] = Hqc::keyGen($p);
        $this->assertSame($p->pkSize, strlen($pk), 'pk size mismatch');
        $this->assertSame($p->skSize, strlen($sk), 'sk size mismatch');

        [$ct, $ss1] = Hqc::encaps($pk, $p);
        $this->assertSame($p->ctSize, strlen($ct), 'ct size mismatch');
        $this->assertSame($p->ssSize, strlen($ss1), 'ss size mismatch');

        $ss2 = Hqc::decaps($sk, $ct, $p);
        $this->assertSame($ss1, $ss2, 'shared secrets do not match');
    }

    public function testKEMDecapsBadCiphertext(): void
    {
        $p = HqcParams::hqc128();
        [$pk, $sk] = Hqc::keyGen($p);
        [$ct, $ss1] = Hqc::encaps($pk, $p);

        // Corrupt ciphertext
        $ct[0] = chr(ord($ct[0]) ^ 0xFF);
        $ct[1] = chr(ord($ct[1]) ^ 0xFF);

        $ss2 = Hqc::decaps($sk, $ct, $p);
        $this->assertNotSame($ss1, $ss2, 'shared secrets should not match with corrupted ciphertext');
    }

    /**
     * @dataProvider paramProvider
     */
    public function testKEMMultipleRoundtrips(HqcParams $p): void
    {
        $trials = 5;
        for ($i = 0; $i < $trials; $i++) {
            [$pk, $sk] = Hqc::keyGen($p);
            [$ct, $ss1] = Hqc::encaps($pk, $p);
            $ss2 = Hqc::decaps($sk, $ct, $p);
            $this->assertSame($ss1, $ss2, "trial $i: shared secrets do not match");
        }
    }

    /**
     * @return array<string, array{0: HqcParams}>
     */
    public static function paramProvider(): array
    {
        return [
            'HQC-128' => [HqcParams::hqc128()],
            'HQC-192' => [HqcParams::hqc192()],
            'HQC-256' => [HqcParams::hqc256()],
        ];
    }
}

File: java/src/test/java/com/pqc/hqc/HqcTest.java
package com.pqc.hqc;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for the HQC KEM implementation.
 */
class HqcTest {

    static Stream<HqcParams> allParams() {
        return Stream.of(HqcParams.allParams());
    }

    // --- GF(256) tests ---

    @Test
    void testGF256Tables() {
        for (int i = 1; i < 256; i++) {
            int logA = GF256.LOG[i];
            int expLogA = GF256.EXP[logA];
            assertEquals(i, expLogA, "exp(log(" + i + ")) should equal " + i);
        }
    }

    @Test
    void testGF256Mul() {
        for (int i = 0; i < 256; i++) {
            assertEquals(i, GF256.mul(1, i), "1 * " + i + " should equal " + i);
        }
        for (int i = 1; i < 256; i++) {
            assertEquals(1, GF256.mul(i, GF256.inv(i)),
                i + " * inv(" + i + ") should equal 1");
        }
    }

    // --- GF(2) vector tests ---

    @Test
    void testGF2VectOps() {
        long[] a = { 0xAAAAL, 0x5555L };
        long[] b = { 0x5555L, 0xAAAAL };
        long[] c = GF2.vectAdd(a, b);
        assertEquals(0xFFFFL, c[0]);
        assertEquals(0xFFFFL, c[1]);

        long[] one = { 1, 0 };
        long[] d = { 0xDEADBEEFCAFEBABEL, 0x1234567890ABCDEFL };
        long[] r = GF2.vectMul(d, one, 128);
        assertEquals(d[0], r[0]);
        assertEquals(d[1], r[1]);
    }

    // --- Reed-Muller tests ---

    @Test
    void testRMEncodeDecodeRoundtrip() {
        for (int msg = 0; msg < 256; msg++) {
            for (int mult : new int[]{ 3, 5 }) {
                int n2 = mult * 128;
                int nWords = (n2 + 63) / 64;
                long[] cw = new long[nWords];
                ReedMuller.encodeInto(cw, msg, 0, mult);
                int decoded = ReedMuller.decode(cw, n2, mult);
                assertEquals(msg, decoded,
                    "RM mult=" + mult + " msg=" + msg + ": got " + decoded);
            }
        }
    }

    // --- Reed-Solomon tests ---

    @ParameterizedTest
    @MethodSource("allParams")
    void testRSEncodeDecodeRoundtrip(HqcParams p) {
        int[] msg = new int[p.k];
        for (int i = 0; i < msg.length; i++) {
            msg[i] = (i + 1) & 0xFF;
        }
        int[] cw = ReedSolomon.encode(msg, p);
        int[] decoded = ReedSolomon.decode(cw, p);
        assertNotNull(decoded, "decode failed on clean codeword");
        assertArrayEquals(msg, decoded);
    }

    @ParameterizedTest
    @MethodSource("allParams")
    void testRSDecodeWithErrors(HqcParams p) {
        int[] msg = new int[p.k];
        for (int i = 0; i < msg.length; i++) {
            msg[i] = (i * 3 + 7) & 0xFF;
        }
        int[] cw = ReedSolomon.encode(msg, p);
        // Introduce delta correctable errors
        for (int i = 0; i < p.delta; i++) {
            cw[i] ^= (i + 1) & 0xFF;
        }
        int[] decoded = ReedSolomon.decode(cw, p);
        assertNotNull(decoded, "decode failed with correctable errors");
        assertArrayEquals(msg, decoded);
    }

    // --- Tensor code tests ---

    @ParameterizedTest
    @MethodSource("allParams")
    void testTensorEncodeDecodeRoundtrip(HqcParams p) {
        byte[] msg = new byte[p.k];
        for (int i = 0; i < msg.length; i++) {
            msg[i] = (byte) (i + 42);
        }
        long[] encoded = TensorCode.encode(msg, p);
        byte[] decoded = TensorCode.decode(encoded, p);
        assertNotNull(decoded, "tensor decode failed");
        assertArrayEquals(msg, decoded);
    }

    // --- KEM roundtrip tests ---

    @ParameterizedTest
    @MethodSource("allParams")
    void testRoundtrip(HqcParams p) {
        byte[][] keys = Hqc.keyGen(p);
        byte[] pk = keys[0];
        byte[] sk = keys[1];

        assertEquals(p.pkSize, pk.length);
        assertEquals(p.skSize, sk.length);

        byte[][] enc = Hqc.encaps(pk, p);
        byte[] ct = enc[0];
        byte[] ss1 = enc[1];

        assertEquals(p.ctSize, ct.length);
        assertEquals(p.ssSize, ss1.length);

        byte[] ss2 = Hqc.decaps(sk, ct, p);
        assertArrayEquals(ss1, ss2, "shared secrets must match");
    }

    @Test
    void testDecapsBadCiphertext() {
        HqcParams p = HqcParams.HQC_128;
        byte[][] keys = Hqc.keyGen(p);
        byte[][] enc = Hqc.encaps(keys[0], p);
        byte[] ct = enc[0];
        byte[] ss1 = enc[1];

        // Corrupt the ciphertext
        ct[0] ^= (byte) 0xFF;
        ct[1] ^= (byte) 0xFF;

        byte[] ss2 = Hqc.decaps(keys[1], ct, p);
        assertFalse(java.util.Arrays.equals(ss1, ss2),
            "shared secrets should not match with corrupted ciphertext");
    }

    @ParameterizedTest
    @MethodSource("allParams")
    void testMultipleRoundtrips(HqcParams p) {
        int trials = 10;
        for (int i = 0; i < trials; i++) {
            byte[][] keys = Hqc.keyGen(p);
            byte[][] enc = Hqc.encaps(keys[0], p);
            byte[] ss2 = Hqc.decaps(keys[1], enc[0], p);
            assertArrayEquals(enc[1], ss2,
                "trial " + i + ": shared secrets do not match");
        }
    }
}
